### ISSUE TITLE
[Core] When an actor gets restarted due to its fault-tolerance behavior, it would be ideal if we could retrieve its logs prior to restart in the UI

### ISSUE THREAD

**Issue by kouroshHakha** [ray-project/ray#43471]
### Description

When an actor gets restarted due to its fault-tolerance behavior, it would be ideal if we could retrieve its logs prior to restart in the UI.

**Comment by ps2181** [comment]
Hi @kouroshHakha,
wanted to see if we can still contribute to this issue.

Looked into the cause — when an actor restarts it gets a new worker process with a new worker_id, so the old log files lose their link to the actor. There's already a TODO in python/ray/dashboard/modules/log/log_manager.py acknowledging this. ActorTableData in GCS only stores the current worker_id with no history of previous ones.

Two approaches seem viable:

- Add a repeated previous_worker_ids field to ActorTableData (GCS proto change) so the dashboard can surface all log files across restarts, indexed by restart attempt using the existing num_restarts counter.
- Mirror how task logs work — they already support attempt_number with byte-offset tracking per attempt. Actors could adopt the same pattern and wire up the attempt_number parameter that already exists in the log API endpoint but is currently only wired for tasks.

Would love to take a stab at this if it's still open for contribution — happy to discuss the preferred approach first.

**Comment by kouroshHakha** [comment]
@ps2181 contribution is always welcome. I think the first approach is sound and light weight. for the dashboard as long as we should the last worker id stack trace or reason for restart, it will be useful enough for debugging etc. 

cc @edoakes for any other inputs.

**Comment by edoakes** [comment]
Agreed, the first approach is preferable. Note that in some cases actors may be restarted many times, so anywhere that we're propagating the previous worker IDs we should have a limit and drop the oldest IDs (perhaps with a warning in the UI if at the limit).

**Comment by ps2181** [comment]
Thanks @kouroshHakha, @edoakes — going with approach #1.

**Scoping the first PR to the plumbing** (dashboard UI as a follow-up so this stays reviewable):

1. **Proto**: add `repeated bytes previous_worker_ids = 39` to `ActorTableData` in `src/ray/protobuf/gcs.proto`.

2. **Writer**: in `GcsActorManager::RestartActor` (`src/ray/gcs/actor/gcs_actor_manager.cc`), append the outgoing `worker_id` to `previous_worker_ids` right before `actor->UpdateAddress(rpc::Address())` clears the address. FIFO-evict oldest when the list exceeds the cap, addressing @edoakes's ask about actors that restart many times.

3. **Cap**: new `RayConfig` knob `actor_previous_worker_ids_limit`, default `10` (open to a different default — let me know). Emits an INFO-level log line the first time we drop an entry per-actor so the "at limit" signal is available for future UI use.

4. **Reader**: update `Log_manager._resolve_actor_filename` (`python/ray/dashboard/modules/log/log_manager.py`) — remove the `TODO(sang)`, iterate `previous_worker_ids` from newest to oldest when the current-worker resolution misses, and return the newest matching file. This makes existing dashboard log links continue to work across restarts without a UI change.

5. **Tests**:
   - **C++**: extend `src/ray/gcs/actor/tests/gcs_actor_manager_test.cc` to verify that after N restarts, `ActorTableData.previous_worker_ids` contains the expected worker IDs in order, and that the list is capped at `actor_previous_worker_ids_limit`.
   - **Python integration**: new test under `python/ray/tests/` that spins a `max_restarts>1` actor, force-kills it a few times, and asserts logs from prior incarnations are retrievable through the log API.

Open to any inputs
